Add support of MacOS

MrOnlineCoder opened the pull request:
This is a one-hell a huge PR that includes 3 features at once:

* Switch to using FetchContent in CMake for downloading dependencies
* Implementation of OpenGL 4.1 RHI backend
* Other fixes needed to get engine and sandbox running on MacOS

I am attaching the screenshot that it's working.
<img width="1920" alt="Screenshot 2024-05-08 at 00 40 51" src="https://github.com/denyskryvytskyi/ElvenEngine/assets/5202478/88972b0e-6ed6-4385-8a86-44e3ee6763d1">

I am sorry in advance for a) big PR b) I maybe written some code or made changes that do not fit your code style and conventions, in this case please let me know what to fix.

This PR is still a WIP/draft because for example I've got the PrimitivesSandbox running, but failed to get MeshModelSandbox up.

I will leave all my questions/concerns that require some discussion from you @denyskryvytskyi 

P.S. even if this does not get merged - it was fun at least as a challenge :D

MrOnlineCoder commented on .gitmodules:
```diff
@@ -1,12 +0,0 @@
-[submodule "Engine/vendor/GLFW"]
```
All dependencies are now fetched via FetchContent, so no submodules are needed

MrOnlineCoder commented on CMakeLists.txt:
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.10)
```
FetchContent is available since cmake 3.11

MrOnlineCoder commented on Engine/CMakeLists.txt:
```diff
@@ -163,6 +171,11 @@ endif()
 set(LIBRARY_NAME ElvenEngine)
 add_library(${LIBRARY_NAME} STATIC ${ENGINE_SOURCES})
 
+target_compile_features(${LIBRARY_NAME} PRIVATE cxx_std_20)
```
For some reason, my clang did not catch using C++ 20 right away for this project, so I forcefully included it here

MrOnlineCoder commented on Engine/CMakeLists.txt:
```diff
@@ -171,102 +184,159 @@ set(VENDOR_PATH ${CMAKE_CURRENT_SOURCE_DIR}/vendor)
+FetchContent_Declare(assimp
+GIT_REPOSITORY "https://github.com/assimp/assimp.git"
+GIT_TAG "v5.3.1")
+set(BUILD_SHARED_LIBS OFF CACHE BOOL "" FORCE)
+set(ASSIMP_BUILD_TESTS OFF CACHE BOOL "" FORCE)
+set(ASSIMP_INJECT_DEBUG_POSTFIX OFF CACHE BOOL "" FORCE)
+set(ASSIMP_INSTALL OFF CACHE BOOL "" FORCE)
+set(ASSIMP_BUILD_ALL_EXPORTERS_BY_DEFAULT OFF CACHE BOOL "" FORCE)
+set(ASSIMP_BUILD_ZLIB OFF CACHE BOOL "" FORCE)
+FetchContent_MakeAvailable(assimp)
```
Configuration for building Assimp. I explicitly disabled building of exporters as not sure you will need them right now. Zlib build also must be disabled to avoid conflicts with our zlib

denyskryvytskyi replied:
@MrOnlineCoder I have an issue while configuring the project. It can't find zlib while configuring assimp. Maybe you could help me.

![image](https://github.com/denyskryvytskyi/ElvenEngine/assets/25298585/58a6e6a4-ab20-496b-bfcc-33d8e69e8b1a)

MrOnlineCoder replied:
Hmm, interesting. I guess we'll have to build zlib strictly before assimp. I will push a fix for it

MrOnlineCoder commented on Engine/CMakeLists.txt:
```diff
@@ -171,102 +184,159 @@ set(VENDOR_PATH ${CMAKE_CURRENT_SOURCE_DIR}/vendor)
+	set(IRRKLANG_BUILD_TESTS OFF CACHE BOOL "" FORCE)
+endif()
+
+add_subdirectory(${VENDOR_PATH}/GLAD)
+add_subdirectory(${VENDOR_PATH}/stb)
```
GLAD and stb are the only libraries that are not FetchContent'd, as they are small enough and filtering out the stb_image from whole repo of stb is a bit of a hassle.

MrOnlineCoder commented on Engine/CMakeLists.txt:
```diff
@@ -171,102 +184,159 @@ set(VENDOR_PATH ${CMAKE_CURRENT_SOURCE_DIR}/vendor)
-    set(ENGINE_INCLUDE_DIRS
-        ${ENGINE_INCLUDE_DIRS}
-        ${VENDOR_PATH}/irrklang/include
+    set(IMGUI_SOURCES
```
imgui does not support cmake for now (it's a long running issue afaik), so we have to assemble it a bit manully

MrOnlineCoder commented on Engine/src/Core/Core.h:
```diff
@@ -15,7 +15,7 @@
 #        define EL_DEBUGBREAK() raise(SIGTRAP)
 #    else
-#        error "Platform doesn't support debugbreak yet!"
+#        define EL_DEBUGBREAK() __builtin_trap()
```
Technically we can replace it with a throw statement

MrOnlineCoder commented on Engine/src/Platform/OpenGL41/OpenGL41Check.h:
```diff
@@ -0,0 +1,19 @@
+    const char* expression);
+}
+
+#ifdef DEBUG_MODE
```
Not sure if PROFILE_MODE is the same as DEBUG_MODE and not sure at all where this macro comes from

denyskryvytskyi replied:
`PROFILE_MODE` macro is used only to enable/disable the internal engine profiler (`Profiler.h`)

MrOnlineCoder replied:
So there is no way currently to know if engine is compiled in debug mode or no? If no, I guess we should add a macro for it - the glChecks would probably be a pointless overhead in release/production

MrOnlineCoder commented on Engine/src/Platform/OpenGL41/OpenGL41Texture.cpp:
```diff
@@ -0,0 +1,105 @@
+    m_internalFormat = format.first;
+    m_dataFormat = format.second;
+
+    // if (info.Multisampled) {
```
Multisampled textures are not supported in OpenGL 4.1 explicitly, so not sure if there is an equivalent.

MrOnlineCoder commented on Engine/src/Platform/OpenGL41/OpenGL41Texture.cpp:
```diff
@@ -0,0 +1,105 @@
+    // } else {
+    glCheck(glGenTextures(1, &m_id));
+    glCheck(glBindTexture(GL_TEXTURE_2D, m_id));
+    if (info.InternalFormat == Texture::InternalFormat::DepthStencil) {
```
not using GL_UNSIGNED_INT_24_8 specifically for GL_DEPTH_STENCIL texture type was problematic and caused errors, had to add a special branch for that

MrOnlineCoder commented on Engine/src/Platform/Windows/WindowsWindow.cpp:
```diff
@@ -102,8 +102,15 @@ void WindowsWindow::Init(const WindowProps& props)
     }
 
+// TODO: avoid using macros here
+#ifdef __APPLE__
+    glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 4);
+    glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 1);
+#else
     glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 4);
     glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 5);
+#endif
+
```
This RHI implementation details leaks into window creation code - the opengl version that we better provide to glfw hints.

What do you think we can do to fix this, assuming other backends like Vulkan (or even DirectX) can be added to the mix with other hints values?

MrOnlineCoder replied:
Moved this to new function "GraphicsContext::PreInit" as glfwWindowHint must be called before the window is created

MrOnlineCoder commented on Engine/src/Renderer/RHI/RendererAPI.cpp:
```diff
@@ -1,9 +1,15 @@
+#include "Platform/OpenGL41/OpenGL41RendererAPI.h"
 
 namespace elv {
+// TODO: change to either runtime check or use more generic factor for checking
+#ifdef __APPLE__
+RendererAPI::API RendererAPI::s_API = RendererAPI::API::OpenGL41;
+#else
 RendererAPI::API RendererAPI::s_API = RendererAPI::API::OpenGL;
+#endif
```
See todo.

denyskryvytskyi replied:
This is enough for now. We can add more specific logic to choose API in the future.

MrOnlineCoder commented on Engine/src/Resources/TextureManager.cpp:
```diff
@@ -68,17 +68,14 @@ SharedPtr<Texture> TextureManager::Load(const std::string& textureName, std::uin
-        }
-        default:
-            EL_CORE_ASSERT(false, "Unknown Renderer API!");
+        SharedPtr<Texture> texture = Texture::Create(
```
Backend selection for texture creation was done inline in texture manager so I had to move it to Texture::Create

MrOnlineCoder commented on Engine/src/Scene/Component.h:
```diff
@@ -74,7 +74,9 @@ class ComponentPool final : public IComponentPool {
     {
         m_entityToComponentIndex.insert({ entity, m_components.size() });
-        m_components.emplace_back(std::forward<Args>(args)...);
+        ComponentType component { std::forward<Args>(args)... };
```
This was very annoying - emplace_back here caused my compiler to spit out unreadable errors complaining about "construct_at" internal method, and according to some info online, this may be a bug in compiler itself or MacOS SDK. So had to use the old way here.

MrOnlineCoder commented on Engine/src/Scene/Systems/Render2dSystem.cpp:
```diff
@@ -43,7 +43,10 @@ void Render2dSystem::OnRender(float dt)
         const auto& entities = m_spritesPool->GetEntities();
         for (auto entity : entities) {
-            sortingInfo.emplace_back(entity, m_trasformsPool->GetComponent(entity).GetPosition().z);
+            SpriteSortingInfo info;
```
Same as above

MrOnlineCoder commented:
Lastly, there are few open points not really related to code:
1) I assume that `Platform/Windows` should've had Win32-specific implementation of window creation and management, however, because you are using GLFW which is cross-platform by itself, I've re-used it for MacOS and didn't find any Windows-specific usage there. Do you think, in order to avoid further confusion, we should a) rename it to something like "GLFW" b) create a complete copy but explicitly name it "MacOS" c) something else? I also suppose it will work for Linux too out-of-the-box

2) Current shaders code uses GLSL version 450 which is, surely, not supported in 4.1 (410). It also uses `binding` feature, which is not supported in older GL. I've adjusted the shaders to get the sandbox running, however not sure how we should maintain it - create a copy of `shaders/` code that will be copied by CMake depending on the platform, or downgrade existing shaders?

3) Sprite fragment shader had hardcoded amount of texture samplers:
```glsl
uniform sampler2D u_Textures[32];
```
However my system supports only 16 of them (this was printed as a specific message from shader compiler), so I had to lower it to 16. Not sure what we should do about it.

4) **Important final note:** this branch may be completely broken for you or your setup (compiler/platform/IDE) so if possible, please test it in an somewhat isolated environment to see if it at least compiles on Windows and is still able to use the newer rendering backend.

denyskryvytskyi commented:
@MrOnlineCoder Great job! I'll review it as soon as possible.

MrOnlineCoder commented:
@denyskryvytskyi I have pushed changed to CMakeLists configuration, and have successfully achieved Sandbox3D app running on Windows 10 (VS 2022), with irrklang too.

During this I have also found that post-build things like copying irrklang dlls are configured in each project separately, i.e. only the Sandbox3D cmakelists are correct now. Can we somehow maybe move this to Engine build script itself, so the engine-projects will only have to copy their assets?


<img width="1073" alt="elven2" src="https://github.com/denyskryvytskyi/ElvenEngine/assets/5202478/e915ffe6-069f-4685-a419-fb89d18eb992">

MrOnlineCoder commented on Engine/CMakeLists.txt:
```diff
@@ -171,102 +184,175 @@ set(VENDOR_PATH ${CMAKE_CURRENT_SOURCE_DIR}/vendor)
+        GIT_TAG "VER-2-13-2"
+)
+
+set(FT_DISABLE_ZLIB ON CACHE BOOL "" FORCE)
+set(FT_DISABLE_BZIP2 ON CACHE BOOL "" FORCE)
+set(FT_DISABLE_PNG ON CACHE BOOL "" FORCE)
+set(FT_DISABLE_HARFBUZZ ON CACHE BOOL "" FORCE)
+set(FT_DISABLE_BROTLI ON CACHE BOOL "" FORCE)
```
These were present in your bat script, so I copied them

denyskryvytskyi replied:
Got it, works for me with the last changes, thanks.
I don't think we need to change post-build commands in this PR. It's already huge, let's avoid complexity increasing.
I'll think about how to reduce irrKlang copying in every project.

MrOnlineCoder commented:
And also important note, right now if you generate a build project via CMake on this branch, you will receive errrors compalining about zlib and zlibstatic. You can ignore them, as the build succeeds even with them.

I am not entirely sure about the nature of this errors and how we can fix them, but due to annoying assimp <-> zlib connection, I would just ignore them for now

denyskryvytskyi commented:
@MrOnlineCoder about zlib. Maybe it's a good idea to remove custom zlib fetching and build the Assimp one. I didn't add it as a separate library in mind.

MrOnlineCoder commented:
Yea, that would be a good option. I wasn't sure if you used zlib somewhere else or if stb_image depended on it. I will remove it then and use the assimp version

denyskryvytskyi commented on Engine/CMakeLists.txt:
```diff
@@ -171,102 +184,175 @@ set(VENDOR_PATH ${CMAKE_CURRENT_SOURCE_DIR}/vendor)
-if (MODULE_EDITOR_ENABLED)
-    find_library(imgui_d NAMES imgui imguid PATHS ${VENDOR_PATH_DEBUG})
-    find_library(imgui NAMES imgui PATHS ${VENDOR_PATH_RELEASE})
+target_include_directories(${LIBRARY_NAME}
+    PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/src
+)
```
You include it twice: once here as public and below as private.

denyskryvytskyi commented:
@MrOnlineCoder a little hint/suggestion: use `set_target_properties(<TARGET> PROPERTIES FOLDER "Libs")` for dependencies to group them in one folder for better usability in IDE.

denyskryvytskyi commented on Engine/src/Core/Window.cpp:
```diff
@@ -1,19 +1,18 @@
 UniquePtr<Window> Window::Create(const WindowProps& props)
 {
-#ifdef EL_PLATFORM_WINDOWS
+    // #ifdef EL_PLATFORM_WINDOWS
```
Please, remove comments if you don't need this code anymore.

denyskryvytskyi commented on Engine/src/Platform/OpenGL41/OpenGL41RenderTarget.cpp:
```diff
@@ -0,0 +1,101 @@
+        info.Multisampled = m_isMultisampled;
+
+        m_depthStencilAttachment = textures::Load(m_size.Width, m_size.Height, info);
+        // glNamedFramebufferTexture(m_id, GL_DEPTH_STENCIL_ATTACHMENT, m_depthStencilAttachment->GetId(), 0);
```
Please, remove the comment.

denyskryvytskyi commented on Engine/src/Platform/OpenGL41/OpenGL41RendererAPI.cpp:
Maybe const.
Also for such complex strings building `fmt::format` library can be used instead of a `+` operator for better readability.

MrOnlineCoder replied:
Good point, forgot about that one

denyskryvytskyi commented:
@MrOnlineCoder Here is my thoughts about different moments:

1. **Platform-specific Window implementation**. You're right, I think we can rename `Platform/WindowsWindow.h` ->`Platform/GLFWWindow.h` 
2. **Shaders**. I'm not sure for now what to do with shaders. Let me think a little bit about it.
3. **About hardcoded amount of samplers in fragment shader for sprite **. Yeah, decreasing it to 16 will be enough for now.
4. **About copying DLLs, assets, and other stuff**. I think we can use copy-paste for sandboxes and game projects for this PR. I'll think about reducing it in the context of a separate task (or you can make suggestions if you will).
5. **GL-specific error on application closing (only for OpenGL 4.1.)**. I have this error only for the OpenGL 4.1. (you may have the same issue). Actually, I think this is because the window context is destroyed before calling destructors of RHI objects like Buffer, Texture, etc. I'm not sure why this exists only for the 4.1 version, but I think it should be fixed with the architecture in mind - OpenGL objects have to call destructors only before window context destroying. I'll think about how to fix it.
![image](https://github.com/denyskryvytskyi/ElvenEngine/assets/25298585/edf3b17d-2454-468c-b44a-15659e06c7a5)
6. Also there is a problem with **MeshModelSandbox: white textures** for OpenGL 4.5. version. According to the log textures aren't loaded at all. Maybe the Assimp texture parsing is broken. Can't say anything, need debugging.

MrOnlineCoder commented:
@denyskryvytskyi thanks for the link on multisample textures, I'll try implementing that (interesting that I didn't find that on official Khronos wiki for some reason), and will test on my Mac machine

I just have two clarifications:
1) Renaming WindowsWindow to GLFWWindow may cause confusions or name collisions, as there is a struct from glfw itself - `GLFWwindow` - the difference is just the uppercase "W" letter, which I am just intuitevly not reallty into. What about `GLFWWindowImpl` or something else? Surely it's more up to you, I am just thinking out loud, as most of Elven code is written inside `namespace elv {...}` blocks, which makes making a typo like that possible. 

2) Do I understand correctly that `Core/Input.h` is implemented as a set of static functions for convenient usage in other classes (components especially), and not as an abstract class that is defined by, let's say `Platform/GLFW`? As I understand in that case, changing then the `GLFWWindow` to some other `Window` implementation would not be enough to cover window/input handling, you would also need to define other implementations for `Input` statics, and also disable the compilation of GLFW-ones via macros for example.
Although I surely doubt the need of writing another window/input handling backend, as glfw is pretty solid and cross-platform solution, and therefore such exception would be fine enough, I am just wondering if you would like to standartize that somehow.

Thanks in advance

denyskryvytskyi commented:
- I've submitted a fix for the "**GL-specific error on application closing**" problem to the main branch 209fb47
- **About GLFW-specific implementations**. Well, a fair point from you. We would need to make all GLFW usage in the separate Window/Input platform-specific code. But, as GLFW is cross-platform and is used for Opengl and will be used for Vulkan, I would not touch this for now. It could be revisited when integrating DirectX at least. So, Just renaming the `WindowsWindow` -> `GlfwWindowImpl` is enough for now. Also, we need to rename `WindowsInput` ->`GlfwInputImpl` I think, but I can do this in the main branch.

MrOnlineCoder commented:
@denyskryvytskyi I've pushed some changes:
* GraphicsContext now accepts window pointer in Init() function instead of constructor
* I've added support for multi sample textures, however, seems I cannot use them as attachment to framebuffer in RenderTarget. I must note that I "suck" in framebuffer APIs, so maybe there is another way for that, please let me know - OpenGL seems to have multiple ways of doing that. So currently I use regular 2D textures in InitAttachments - this way the sandbox app works
* Renamed WindowsWindow and WindowsInput to GLFWWindowImpl and GLFWInputImpl respectively

denyskryvytskyi commented on Engine/src/Platform/OpenGL41/OpenGL41RenderTarget.cpp:
```diff
@@ -0,0 +1,99 @@
+            glFramebufferTexture2D(
+                GL_FRAMEBUFFER,
+                GL_COLOR_ATTACHMENT0,
+                GL_TEXTURE_2D,
```
Try to write `m_isMultisampled ? GL_TEXTURE_2D_MULTISAMPLE : GL_TEXTURE_2D,` here and set `info.Multisample`. It works for me for OpenGL41 API. If it doesn't work on Mac, then I think we can skip MSAA for opengl41.

MrOnlineCoder replied:
AFAIK, I have tried exactly that and didn't succeed. Although the pre-debugcallback error messages are not very descriptive (and not always documented), so maybe I missed something - I will try again, and if not, no MSAA would be available for opengl41 for now.

denyskryvytskyi commented:
Thoughts about shaders: it's better to make a separate folder for older version of shaders and adapt code for that.

MrOnlineCoder commented:
@denyskryvytskyi 

> Because StaticMeshComponent is subscribed on the model-loaded event in the constructor pushing a copy of the created object invalidates this pointer in the event handler. I'm not sure how to fix it for now, but creating a copy is not a good idea here.

I am a bit weak on newer C++ features and typing system, so I really can't help there directly. I will try researching for a way to get this compilation error fixed on MacOS (maybe by changing the build toolchain)

> Thoughts about shaders: it's better to make a separate folder for older version of shaders and adapt code for that.
Would you like me to try doing that in current PR, or do you have any specific vision on the implementation itself?

denyskryvytskyi commented:
Well, I don't think shaders duplication/adaptation has any meaning out of this PR.
There are two options you could do:
1. **No duplication**. Remove the `binding` feature from the shaders (and adapt C++ code, we need just bind these textures), downgrade the `version`, and test it. If it will work that is enough.
2. **Duplication [Not desirable because for every new shader or refactoring one needs us to do the same for the old one]** if the first point doesn't work and other new features aren't supported. In this case, we need to take the following actions:
    - Make another folder in assets like `shadersGL41`
    - adapt shaders for OpenGL 4.1
    - Copy the appropriate folder to the binary `shader` folder according to the platform. I think this is enough for now.
    

Anyway, both points involve C++ code adaptation, so the first one is better if It doesn't require a lot of downgrade.
